StoredProcedureEFCore.Tests.SqlServer/Program.cs

```diff
@@ -87,16 +90,160 @@ await ctx.LoadStoredProc("dbo.OutputNullable")
+
+            await TestCancelBeforeResultset(ctx, 7, async (r, ct) => await r.ToLookupAsync<Int64, Model>(m => m.Id, ct));
+            await TestCancelAfterResultset(ctx, 8, async (r, ct) => await r.ToLookupAsync<Int64, Model>(m => m.Id, ct));
+
+            await TestCancelBeforeResultset(ctx, 9, async (r, ct) => await r.ToSetAsync<Int64>(ct));
+            await TestCancelAfterResultset(ctx, 10, async (r, ct) => await r.ToSetAsync<Int64>(ct));
+
+            await TestCancelBeforeResultset(ctx, 11, async (r, ct) => await r.FirstAsync<Model>(ct));
+            await TestCancelAfterResultset(ctx, 12, async (r, ct) => await r.FirstAsync<Model>(ct));
+
+            await TestCancelBeforeResultset(ctx, 13, async (r, ct) => await r.SingleAsync<Model>(ct), 1);
+            await TestCancelAfterResultset(ctx, 14, async (r, ct) => await r.SingleAsync<Model>(ct), 1);
+
+            await TestCancelBeforeResultset(ctx, 15, async (r, ct) => await r.SingleOrDefaultAsync<Model>(ct), 1);
+            await TestCancelAfterResultset(ctx, 16, async (r, ct) => await r.SingleOrDefaultAsync<Model>(ct), 1);
+        }
+
+#pragma warning disable AsyncFixer04 // A disposable object used in a fire & forget async call
+        /// <summary>
+        /// Test async cancellation where the sproc is cancelled before it has a chance to produce a resultset.
+        /// </summary>
+        static async Task TestCancelBeforeResultset(DbContext ctx, int callId, Func<DbDataReader, CancellationToken, Task> action, Int64 limit = 9223372036854775807)
+        {
+            var stopWatch = new Stopwatch();
+            stopWatch.Start();
+
+            try
+            {
+                Console.WriteLine("\n====================================");
+                Console.WriteLine($"Starting long-running call #{callId}...");
+
+                using (var cts = new CancellationTokenSource())
+                {
+                    var sprocTask = ctx.LoadStoredProc("dbo.ListAll")
+                        .AddParam("limit", limit)
+                        .AddParam("delay_in_seconds_before_resultset", 10)
+                        .ExecAsync(async r => await action(r, cts.Token), cts.Token);
+
+                    // Cancel the task after one second.
+                    var cancelTask = CancelTokenAsync(callId, 1000, cts);
+
+                    // Wait for the sproc and the cancellation task to complete.
+                    await Task.WhenAll(sprocTask, cancelTask);
+
+                    // If the task was cancelled properly, we should not be here.
+                    throw new InvalidOperationException("Task was supposed to be cancelled.");
+                }
+            }
+            catch (SqlException e)
+            {
+                // If the task was cancelled properly, SqlClient's cmd.ExecuteAsync should throw SqlException with "Operation cancelled by user" in message.
+                if (!e.Message.Contains("Operation cancelled by user"))

```

Comment:
I don't like at all comparing exception's message. Wouldn't `e is OperationCanceledException` do the same ?

Reply:
Not really.
The issue here is that cancelling SqlCommand.ExecuteAsync through a cancellation token does not throw OperationCancelledException (as one might expect), but rather the generic SqlException with a message which contains the snippet "Operation cancelled by user".
I believe this is the result of SQL Server cancellation semantics existing prior to the introduction of async/await and the cancellation token.